azurelinuxagent/common/protocol/imds.py

```diff
@@ -254,38 +261,69 @@ def _http_get(self, endpoint, resource_path, headers):
+            response: IMDS response on IMDS_RESPONSE_SUCCESS, failure message otherwise
+        """
+        try:
+            resp = self._http_get(endpoint=endpoint, resource_path=resource_path, headers=headers)
+        except ResourceGoneError:
+            return IMDS_INTERNAL_SERVER_ERROR, "IMDS error in /metadata/{0}: HTTP Failed with Status Code 410: Gone".format(resource_path)
+        except HttpError as e:
+            msg = str(e)
+            if self._regex_throttled.match(msg):
+                return IMDS_RESPONSE_ERROR, "IMDS error in /metadata/{0}: Throttled".format(resource_path)
+            if self._regex_ioerror.match(msg):
+                logger.periodic_warn(logger.EVERY_FIFTEEN_MINUTES,
+                                     "[PERIODIC] [IMDS_CONNECTION_ERROR] Unable to connect to IMDS endpoint {0}".format(endpoint))
+                return IMDS_CONNECTION_ERROR, "IMDS error in /metadata/{0}: Unable to connect to endpoint".format(resource_path)
+            return IMDS_INTERNAL_SERVER_ERROR, "IMDS error in /metadata/{0}: {1}".format(resource_path, msg)
+
+        if resp.status >= 500:
+            return IMDS_INTERNAL_SERVER_ERROR, "IMDS error in /metadata/{0}: {1}".format(
+                                               resource_path, restutil.read_response_error(resp))
+
+        if restutil.request_failed(resp):
+            return IMDS_RESPONSE_ERROR, "IMDS error in /metadata/{0}: {1}".format(
+                                        resource_path, restutil.read_response_error(resp))
+
+        return IMDS_RESPONSE_SUCCESS, resp.read()
+
     def get_metadata(self, resource_path, is_health):
         """
         Get metadata from IMDS, falling back to Wireserver endpoint if necessary.
 
         :param str resource_path: path of IMDS resource
         :param bool is_health: True if for health/heartbeat, False otherwise
-        :return: Tuple<is_request_success:bool, response:str>
-            is_request_success: True when connection succeeds, False otherwise
-            response: response from IMDS on request success, failure message otherwise
+        :return: instance of MetadataResult
+        :rtype: MetadataResult
         """
         headers = self._health_headers if is_health else self._headers
         endpoint = IMDS_ENDPOINT
-        try:
-            resp = self._http_get(endpoint=endpoint, resource_path=resource_path, headers=headers)
-        except HttpError as e:
+
+        status, resp = self._get_metadata_from_endpoint(endpoint, resource_path, headers)
+        if status == IMDS_CONNECTION_ERROR:
             logger.periodic_warn(logger.EVERY_FIFTEEN_MINUTES,
-                                 "[PERIODIC] Unable to connect to primary IMDS endpoint {0}".format(endpoint))
-            if not self._regex_imds_ioerror.match(str(e)):
-                raise
+                                 "[PERIODIC] [IMDS_CONNECTION_ERROR] Unable to connect to primary IMDS endpoint {0}".format(endpoint))
```

Comment:
Wouldn't this message be logged already in L285?

Reply:
Yes and no. This message adds "primary" or "secondary" to the log message, but I agree that this is not actually important since we already log the endpoint. I'll remove these as duplicate.